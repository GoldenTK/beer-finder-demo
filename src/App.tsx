import React, { useState } from 'react'
import { defaultsDeep } from 'lodash/fp'
import { ThemeProvider } from '@material-ui/styles'
import { createMuiTheme } from '@material-ui/core/styles'
import './App.css'
import { CssBaseline } from '@material-ui/core'
import { Switch, Route, Redirect, BrowserRouter as Router } from 'react-router-dom'
import Home from './Home'
import Details from './Details'
import { RecursivePartial } from './types'

const theme = createMuiTheme({
  palette: {
    type: 'dark',
  }
})

theme.spacing(2)

const redirectToHome = () =>
  <Redirect to="/" />

export interface FilterParams {
  name: string,
  malt: string
}

export interface GlobalState {
  page: number,
  rowsPerPage: number,
  filter: FilterParams
}

export type PartialGlobalState = RecursivePartial<GlobalState>
export const initialState = {
  page: 0,
  rowsPerPage: 5,
  filter: {
    name: '',
    malt: ''
  }
}

const App: React.FC = () => {
  const [globalState, setGlobalState] = useState<GlobalState>(initialState)

  const handleStateChange = (newState: PartialGlobalState) =>
    setGlobalState(defaultsDeep(globalState, newState))

  return (
    <ThemeProvider theme={theme}>
      <Router>
        <CssBaseline />
        <Switch>
          <Route exact path="/">
            <Home
              state={globalState}
              onStateChange={handleStateChange}
            />
          </Route>
          <Route path="/beer/:beerId" component={Details} />
          <Route path="*" render={redirectToHome} />
        </Switch>
      </Router>
    </ThemeProvider>
  )
}

export default App
