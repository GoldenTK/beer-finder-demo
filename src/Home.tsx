import React, { FC, useState, useCallback } from 'react'
import { debounce } from 'lodash/fp'
import { makeStyles, createStyles, Typography, TextField } from '@material-ui/core'
import BeersList from './BeersList'
import { TableOptions } from './App'

const useStyles = makeStyles(() =>
  createStyles({
    filterGroup: {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'space-evenly'
    },
    formControl: {
      minWidth: 120,
    },
  }),
)

const Home: FC<{
  options: TableOptions,
  onOptionsChange: (newOptions: any) => void
}> = ({ options, onOptionsChange }) => {
  const classes = useStyles()
  const [filter, setFilter] = useState({ name: options.filter.name, malt: options.filter.malt })

  const memoizedFilterUpdate = useCallback(debounce(1000, (name: string, malt: string) =>
    setFilter({ name, malt })), [])

  const handleBeerNameChange = (event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    onOptionsChange({ filter: { name: event.target.value }, page: 0 })
    memoizedFilterUpdate(event.target.value, options.filter.malt)
  }

  const handleMaltChange = (event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    onOptionsChange({ filter: { malt: event.target.value }, page: 0 })
    memoizedFilterUpdate(options.filter.name, event.target.value)
  }

  return (
    <>
      <div className={classes.filterGroup}>
        <Typography variant="subtitle1">Filter: </Typography>
        <TextField label="Name" value={options.filter.name} onChange={handleBeerNameChange} />
        <TextField label="Malt" value={options.filter.malt} onChange={handleMaltChange} />
      </div>
      <BeersList
        filter={filter}
        page={options.page}
        rowsPerPage={options.rowsPerPage}
        onOptionsChange={onOptionsChange}
      />
    </>
  )
}

export default Home
