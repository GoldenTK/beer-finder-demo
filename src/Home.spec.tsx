import { spy } from 'sinon'
import { identity } from 'lodash/fp'
import { shallow } from 'enzyme'
import React from 'react'
import Home from './Home'
import { initialState } from './App'
import { TextField } from '@material-ui/core'

it('renders correct structure', () => {
  const home = shallow(<Home state={initialState} onStateChange={identity} />)
  expect(home).toMatchSnapshot()
})

it('should call state update on beer name change', () => {
  const stateChangeSpy = spy()
  const home = shallow(<Home state={initialState} onStateChange={stateChangeSpy} />)

  home.find(TextField).at(0).simulate('change', {
    target: {
      value: 'value'
    }
  })

  stateChangeSpy.calledWith({ filter: { name: 'value' }, page: 0 })
})

it('should call state update on malt change', () => {
  const stateChangeSpy = spy()
  const home = shallow(<Home state={initialState} onStateChange={stateChangeSpy} />)

  home.find(TextField).at(0).simulate('change', {
    target: {
      value: 'value'
    }
  })

  stateChangeSpy.calledWith({ filter: { malt: 'value' }, page: 0 })
})
