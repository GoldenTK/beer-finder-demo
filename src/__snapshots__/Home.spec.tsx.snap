// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correct structure 1`] = `ShallowWrapper {}`;
